lib/utils/findPort.js

```diff
@@ -1,39 +1,36 @@
 'use strict';
 
-const pRetry = require('p-retry');
 const portfinder = require('portfinder');
 const defaultPort = require('./defaultPort');
-const defaultTo = require('./defaultTo');
 const tryParseInt = require('./tryParseInt');
 
-function runPortFinder() {
-  return new Promise((resolve, reject) => {
-    portfinder.basePort = defaultPort;
-    portfinder.getPort((error, port) => {
-      if (error) {
-        return reject(error);
-      }
-
-      return resolve(port);
-    });
-  });
-}
-
-function findPort(port) {
-  if (port) {
-    return Promise.resolve(port);
+async function findPort(options) {
+  if (tryParseInt(options)) {
+    return tryParseInt(options);
   }
 
-  // Try to find unused port and listen on it for 3 times,
-  // if port is not specified in options.
-  // Because NaN == null is false, defaultTo fails if parseInt returns NaN
-  // so the tryParseInt function is introduced to handle NaN
-  const defaultPortRetry = defaultTo(
-    tryParseInt(process.env.DEFAULT_PORT_RETRY),
-    3
-  );
+  const defaultRetry = tryParseInt(process.env.DEFAULT_PORT_RETRY) || 3;
+
+  portfinder.basePort = defaultPort;
+  portfinder.highestPort = portfinder.basePort + defaultRetry;
+
+  if (Object.prototype.toString.call(options) === '[object Object]') {
+    const basePort = tryParseInt(options.basePort);
+    const highestPort = tryParseInt(options.highestPort);
+    const retry = tryParseInt(options.retry);
+
+    if (basePort) {
+      portfinder.basePort = basePort;
+    }
+
+    if (retry) {
+      portfinder.highestPort = portfinder.basePort + retry;
+    } else if (highestPort) {
+      portfinder.highestPort = options.highestPort;
+    }
+  }
 
-  return pRetry(runPortFinder, { retries: defaultPortRetry });
+  return portfinder.getPortPromise();
```

Comment:
I think here breaking change, when i use `8080` port and when port is busy next port should be `8081`, then `8082` and continue, look `basePort` and `highestPort` just get random port from range.

Reply:
Not breaking change. See https://github.com/http-party/node-portfinder/blob/master/lib/portfinder.js#L357-L359

Reply:
We need avoid this breaking, because it is good logic when port is unavailable we get next port instead random, just need test when we use `port` as number/string/null and port unavailable we get next port instead random from range

Reply:
How to resolve this situation?

```
{
  basePort: 8080,
  highestPort: 9000
}
```

and default retry is 3.

1. 8080 <- used
2. 8087 <- used
3. 9000 <- used

in this case, we'll finish searching even if we can't find an empty port?
Random retry is very unstable.

I think if highestPort is set we might be better to ignore retry.

Reply:
We should avoid `retry` option for this case, we already say range of ports, retry option is needed only when we use `port` as number/null or string

Reply:
if `port` is set, it will go to https://github.com/webpack/webpack-dev-server/pull/2196/files#diff-487ca80b65e47f5165c8419f1664af7cR8 and finish. And if `port` is `null` or `undefined`, this code has already set the highest port. https://github.com/webpack/webpack-dev-server/pull/2196/files#diff-487ca80b65e47f5165c8419f1664af7cR15

Reply:
I want to say what we really do not need `retry` option for `object` notation, because it doesn't have sense, if i write:
```
{
  basePort: 8080,
  highestPort: 9000
}
```

I want any port from this range and the retry option in this case is misleading, what `retry` means in this case? Try to get busy multiple times busy port (for example 8080 is busy and we try to get 8080 again and again) or try to get other free port (for example 8080 is busy and we try to get 8081 then 8082)?

To be honesty retry env variable is very misleading also, maybe we need rename this option here too

thoughts?

Reply:
> I want any port from this range and the retry option in this case is misleading,  

Yes, I think so too. Actually, I confused when I was implementing this feature.

In the first place we offer the highestPort option, so is this retry option still necessary?(just providing the default value(3) or rename `DEFAULT_PORT_RETRY` ...)

Reply:
Give me time to think about it, I'm a little busy today, so I'll write my thoughts tomorrow :+1:

Reply:
Yes, sure. Thank you for your review!